internal/driver/onvifdiscovery.go

```diff
@@ -0,0 +1,204 @@
+				SecretPath:         d.config.DefaultSecretPath,
+				EndpointRefAddress: endpointRefAddr,
+			},
+		},
+	}
+
+	devInfo, edgexErr := d.getDeviceInformation(device)
+	if edgexErr != nil {
+		// try again using the endpointRefAddr as the SecretPath. the reason for this is that
+		// the user may have pre-filled the secret store with per-device credentials based on the endpointRefAddr.
+		device.Protocols[OnvifProtocol][SecretPath] = endpointRefAddr
+		devInfo, edgexErr = d.getDeviceInformation(device)
+	}
+
+	var discovered sdkModel.DiscoveredDevice
+	if edgexErr != nil {
+		d.lc.Warnf("failed to get the device information for the camera %s, %v", endpointRefAddr, edgexErr)
+		discovered = sdkModel.DiscoveredDevice{
+			Name:        endpointRefAddr,
+			Protocols:   device.Protocols,
+			Description: "Auto discovered Onvif camera",
+			Labels:      []string{"auto-discovery"},
+		}
+		d.lc.Debugf("Discovered unknown camera '%s' from the address '%s'", discovered.Name, xaddr)
+	} else {
+		device.Protocols[OnvifProtocol][Manufacturer] = devInfo.Manufacturer
+		device.Protocols[OnvifProtocol][Model] = devInfo.Model
+		device.Protocols[OnvifProtocol][FirmwareVersion] = devInfo.FirmwareVersion
+		device.Protocols[OnvifProtocol][SerialNumber] = devInfo.SerialNumber
+		device.Protocols[OnvifProtocol][HardwareId] = devInfo.HardwareId
+
+		// Spaces are not allowed in the device name
+		deviceName := fmt.Sprintf("%s-%s-%s",
+			strings.ReplaceAll(devInfo.Manufacturer, " ", "-"),
+			strings.ReplaceAll(devInfo.Model, " ", "-"),
+			endpointRefAddr)
+
+		discovered = sdkModel.DiscoveredDevice{
+			Name:        deviceName,
+			Protocols:   device.Protocols,
+			Description: fmt.Sprintf("%s %s Camera", devInfo.Manufacturer, devInfo.Model),
+			Labels:      []string{"auto-discovery", devInfo.Manufacturer, devInfo.Model},
+		}
+		d.lc.Debugf("Discovered camera '%s' from the address '%s'", discovered.Name, xaddr)
+	}
+	return discovered, nil
+}
+
+// mapProbeResults converts a slice of discovered onvif.Device into the generic
+// netscan.ProbeResult.
+func mapProbeResults(host, port string, devices []onvif.Device) (res []netscan.ProbeResult) {
+	for _, device := range devices {
+		res = append(res, netscan.ProbeResult{
+			Host: host,
+			Port: port,
+			Data: device,
+		})
+	}
+	return res
+}
+
+// executeRawProbe essentially performs a UDP unicast ws-discovery probe by sending the
+// probe message directly over the connection and listening for any responses. Those
+// responses are then converted into a slice of onvif.Device.
+func executeRawProbe(conn net.Conn, params netscan.Params) ([]onvif.Device, error) {
+	probeSOAP := wsdiscovery.BuildProbeMessage(uuid.Must(uuid.NewV4()).String(), nil, nil,
+		map[string]string{"dn": "http://www.onvif.org/ver10/network/wsdl"})
+
+	addr := conn.RemoteAddr().String()
+	if err := conn.SetDeadline(time.Now().Add(params.Timeout)); err != nil {
+		return nil, errors.Wrapf(err, "%s: failed to set read/write deadline", addr)
```

Comment:
Can we use EdgeX's errors package instead of pkg/errors?

Reply:
yeah, let me convert this.